File: test/dynamics_vibrations_tests.f90
module dynamics_vibrations_tests
    use iso_fortran_env
    use dynamics
    use fortran_test_helper
    implicit none
contains
! ------------------------------------------------------------------------------
function test_q_factor() result(rst)
    ! Arguments
    logical :: rst

    ! Local Variables
    real(real64) :: q, ans, zeta

    ! Initialization
    rst = .true.
    call random_number(zeta)
    ans = 1.0d0 / (2.0d0 * zeta)

    ! Test
    q = q_factor(zeta)
    if (.not.assert(q, ans)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_q_factor -1"
    end if
end function

! ------------------------------------------------------------------------------
function test_bandwidth() result(rst)
    ! Arguments
    logical :: rst

    ! Local Variables
    real(real64) :: df, ans, fn, zeta

    ! Initialization
    rst = .true.
    call random_number(zeta)
    call random_number(fn)
    ans = 2.0d0 * zeta * fn

    ! Test
    df = estimate_bandwidth(fn, zeta)
    if (.not.assert(df, ans)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_bandwidth -1"
    end if
end function

! ------------------------------------------------------------------------------
function test_log_decrement() result(rst)
    ! Arguments
    logical :: rst

    ! Variables
    integer(int32), parameter :: n = 2
    real(real64) :: x1, x2, ans, delta

    ! Initialization
    rst = .true.
    call random_number(x1)
    x2 = 0.8d0 * x1
    ans = log(x1 / x2) / n

    ! Test
    delta = logarithmic_decrement(x1, x2, n)
    if (.not.assert(delta, ans)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_log_decrement -1"
    end if
end function

! ------------------------------------------------------------------------------
function test_damping_from_decrement() result(rst)
    ! Arguments
    logical :: rst

    ! Variables
    real(real64), parameter :: pi = 2.0d0 * acos(0.0d0)
    real(real64) :: delta, zeta, ans

    ! Initialization
    rst = .true.
    call random_number(delta)
    ans = delta / sqrt(4.0d0 * pi**2 + delta**2)

    ! Test
    zeta = damping_from_log_decrement(delta)
    if (.not.assert(zeta, ans)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_damping_from_decrement -1"
    end if
end function

! ------------------------------------------------------------------------------
function test_find_free_rsp_props() result(rst)
    ! Arguments
    logical :: rst

    ! Parameters
    real(real64), parameter :: fn = 1.0d1
    real(real64), parameter :: zeta = 1.0d-1
    real(real64), parameter :: xo = 0.0d0
    real(real64), parameter :: vo = 1.0d0
    real(real64), parameter :: pi = 2.0d0 * acos(0.0d0)
    real(real64), parameter :: dt = 1.0d-3
    integer(int32), parameter :: n = 1000

    ! Variables
    integer(int32) :: i
    real(real64) :: t(n), x(n), wn, wd, A, B, fnx, delta, zx, tol

    ! Initialization
    rst = .true.
    t = (/ (i * dt, i = 0, n - 1) /)
    wn = 2.0d0 * pi * fn
    wd = wn * sqrt(1.0d0 - zeta**2)
    A = (vo + zeta * wn * xo) / wd
    B = xo
    x = exp(-zeta * wn * t) * (A * sin(wd * t) + B * cos(wd * t))
    tol = 1.0d-3

    ! Test
    call find_free_response_properties(t, x, delta, fnx)
    zx = damping_from_log_decrement(delta)

    if (.not.assert(zx, zeta, tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_find_free_rsp_props -1"
    end if

    tol = abs(wn - wd)
    if (.not.assert(fnx, fn, tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_find_free_rsp_props -2"
    end if
end function

! ------------------------------------------------------------------------------
function test_rise_time() result(rst)
    ! Arguments
    logical :: rst

    ! Parameters
    real(real64), parameter :: pi = 2.0d0 * acos(0.0d0)

    ! Local Variables
    real(real64) :: wn, zeta, ans, tr

    ! Initialization
    rst = .true.
    call random_number(wn)
    call random_number(zeta)
    ans = (1.0d0 / (wn * sqrt(1.0d0 - zeta**2))) * (pi - &
        atan(sqrt(1.0d0 - zeta**2) / zeta))

    ! Test
    tr = rise_time(wn, zeta)
    if (.not.assert(ans, tr)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_rise_time -1"
    end if
end function

! ------------------------------------------------------------------------------
end module

File: test/c/dynamics_vibrations_tests.c
#include "dynamics.h"
#include "dynamics_vibrations_tests.h"
#include "dynamics_c_test_helper.h"
#include <math.h>
#include <stdio.h>

bool c_test_q_factor()
{
    // Local Variables
    bool rst;
    double q, ans;
    const double zeta = 0.25;
    const double tol = 1.0e-8;

    // Initialization
    rst = true;
    ans = 1.0 / (2.0 * zeta);

    // Test
    q = c_q_factor(zeta);
    if (fabs(q - ans) > tol) 
    {
        rst = false;
        printf("TEST FAILED: c_test_q_factor -1\n");
    }

    // End
    return rst;
}


bool c_test_bandwidth()
{
    // Local Variables
    bool rst;
    double df, ans;
    const double fn = 1.0e3;
    const double zeta = 1.0e-1;
    const double tol = 1.0e-8;

    // Initialization
    rst = true;
    ans = 2.0 * zeta * fn;

    // Test
    df = c_estimate_bandwidth(fn, zeta);
    if (fabs(df - ans) > tol)
    {
        rst = false;
        printf("TEST FAILED: c_test_bandwidth -1\n");
    }

    // End
    return rst;
}


bool c_test_log_decrement()
{
    // Local Variables
    bool rst;
    double x1, x2, ans, delta;
    const int n = 2;
    const double tol = 1.0e-8;

    // Initialization
    rst = true;
    x1 = 2.3;
    x2 = 0.8 * x1;
    ans = log(x1 / x2) / n;

    // Test
    delta = c_logarithmic_decrement(x1, x2, n);
    if (fabs(delta - ans) > tol)
    {
        rst = false;
        printf("TEST FAILED: c_test_log_decrement -1\n");
    }

    // End
    return rst;
}


bool c_test_damping_from_decrement()
{
    // Local Variables;
    bool rst;
    double pi, delta, zeta, ans;
    const double tol = 1.0e-8;

    // Initialization
    rst = true;
    pi = 2.0 * acos(0.0);
    delta = 0.5;
    ans = delta / sqrt(4.0 * pi * pi + delta * delta);

    // Test
    zeta = c_damping_from_log_decrement(delta);
    if (fabs(zeta - ans) > tol)
    {
        rst = false;
        printf("TEST FAILED: c_test_damping_from_decrement -1\n");
    }

    // End
    return rst;
}


bool c_test_find_free_rsp_props()
{
    // Local Variables
    bool rst;
    const double fn = 1.0e1;
    const double zeta = 1.0e-1;
    const double xo = 0.0;
    const double vo = 1.0;
    const double dt = 1.0e-3;
    const int n = 1000;
    const double tol = 1.0e-2;
    const double s = 1.0e-2;
    const int np = 2;
    int i;
    double pi, t[n], x[n], wn, wd, A, B, fnx, delta, zx, x1, x2, t1, t2;

    // Initialization
    rst = true;
    pi = 2.0 * acos(0.0);
    wn = 2.0 * pi * fn;
    wd = wn * sqrt(1.0 - zeta * zeta);
    A = (vo + zeta * wn * xo) / wd;
    B = xo;
    for (i = 0; i < n; ++i)
    {
        t[i] = i * dt;
        x[i] = exp(-zeta * wn * t[i]) * (A * sin(wd * t[i]) + B * cos(wd * t[i]));
    }

    // Test
    c_find_free_response_properties(n, t, x, s, np, &delta, &fnx, &x1, &x2, 
        &t1, &t2);
    zx = c_damping_from_log_decrement(delta);

    if (fabs(zx - zeta) > tol)
    {
        rst = false;
        printf("TEST FAILED: c_test_find_free_rsp_props -1\n");
    }

    if (fabs(fnx - fn) > tol)
    {
        rst = false;
        printf("TEST FAILED: c_test_find_free_rsp_props -2\n");
    }

    // End
    return rst;
}


bool c_test_rise_time()
{
    // Local Variables
    bool rst;
    const double tol = 1.0e-8;
    const double wn = 1.52e3;
    const double zeta = 0.35;
    double pi, ans, tr, arg;

    // Initialization
    rst = true;
    pi = 2.0 * acos(0.0);
    arg = sqrt(1.0 - zeta * zeta);
    ans = (1.0 / (wn * arg)) * (pi - atan(arg / zeta));

    // Test
    tr = c_rise_time(wn, zeta);
    if (fabs(ans - tr) > tol)
    {
        rst = false;
        printf("TEST FAILED: c_test_rise_time -1\n");
    }

    // End
    return rst;
}


bool c_test_step_response()
{
    // Local Variables
    bool rst;
    const int n = 1000;
    const double dt = 1.0e-3;
    const double fn = 1.5e1;
    const double zeta = 3.5e-1;
    const double xs = 2.0;
    const double tol = 1.0e-8;
    int i;
    double t[n], x[n], ans[n], wn, wd, A, pi;

    // Initialization
    rst = true;
    wn = 2.0 * pi * fn;
    wd = wn * sqrt(1.0 - zeta * zeta);
    A = zeta * wn / wd;
    for (i = 0; i < n; ++i)
    {
        t[i] = i * dt;
        ans[i] = xs * (1.0 - exp(-zeta * wn * t[i]) * (A * sin(wd * t[i]) + cos(wd * t[i])));
    }

    // Test
    c_evaluate_step_response(n, wn, zeta, xs, t, x);
    if (!compare_arrays(n, ans, x, tol))
    {
        rst = false;
        printf("TEST FAILED: c_test_step_response -1\n");
    }

    // End
    return rst;
}


bool c_test_settling_amplitude()
{
    // Local Variables
    bool rst;
    int i;
    const int n = 1000;
    const double tol = 5.0e-2;
    const double dt = 1.0e-3;
    const double fn = 1.3e1;
    const double zeta = 0.25;
    const double xs = 2.0;
    double pi, t[n], x[n], xf, wn;

    // Initialization
    rst = true;
    pi = 2.0 * acos(0.0);
    wn = 2.0 * pi * fn;
    for (i = 0; i < n; ++i)
    {
        t[i] = i * dt;
    }
    c_evaluate_step_response(n, wn, zeta, xs, t, x);

    // Test
    xf = c_find_settling_amplitude(n, x);
    if (fabs(xf - xs) > tol)
    {
        rst = false;
        printf("TEST FAILED: c_test_settling_amplitude -1\n");
    }

    // End
    return rst;
}


bool c_test_damping_from_overshoot()
{
    // Local Variables
    bool rst;
    const int n = 1000;
    const double tol = 0.1;
    const double dt = 1.0e-3;
    const double fn = 1.2e1;
    const double zeta = 0.82;
    const double xs = 2.0;
    int i;
    double pi, wn, zx, t[n], x[n];

    // Initialization
    rst = true;
    pi = 2.0 * acos(0.0);
    wn = 2.0 * pi * fn;
    for (i = 0; i < n; ++i)
    {
        t[i] = i * dt;
    }
    c_evaluate_step_response(n, wn, zeta, xs, t, x);

    // Test
    zx = c_damping_from_fractional_overshoot(n, x);
    if (fabs(zx - zeta) > tol)
    {
        rst = false;
        printf("TEST FAILED: c_test_damping_from_overshoot -1\n");
    }

    // End
    return rst;
}
